### PR TITLE
JAMES-3167 Reactify MailboxMapper

### PR DESCRIPTION
**Pull request by chibenwa**
Extract MailboxMapper related changes from https://github.com/linagora/james-project/pull/3281/

### CODE REVIEW

**Comment by chibenwa** [comment]
```
[2addfd023d3f559fca7e8eb836def53b4ab0955b] [ERROR] Failures: 
[2addfd023d3f559fca7e8eb836def53b4ab0955b] [ERROR]   StoreRightManagerTest.hasRightShouldThrowMailboxNotFoundExceptionWhenMailboxDoesNotExist:94 
[2addfd023d3f559fca7e8eb836def53b4ab0955b] Expecting a throwable with cause being an instance of:
[2addfd023d3f559fca7e8eb836def53b4ab0955b]  <org.apache.james.mailbox.exception.MailboxNotFoundException>
[2addfd023d3f559fca7e8eb836def53b4ab0955b] but current throwable has no cause.
```

File: mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerTest.java [hunk 1]
```diff
@@ -558,7 +558,7 @@ void updateAnnotationsShouldThrowExceptionIfMailboxDoesNotExist() throws Excepti
             MailboxPath inbox = MailboxPath.inbox(session);
 
             assertThatThrownBy(() -> mailboxManager.updateAnnotations(inbox, session, ImmutableList.of(privateAnnotation)))
-                .isInstanceOf(MailboxNotFoundException.class);
+                .hasCauseInstanceOf(MailboxNotFoundException.class);
```
Comment (rouazana) [inline mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerTest.java:561]: I did not review everything but that's my first question to try to understand this changeset.

In this commit MailboxAnnotationManager did not changed and is still throwing "MailboxException". So why do you need this change? Did the API changed in a non documented way?

Comment (chibenwa) [reply mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerTest.java:561]: Well that's likely something that will break and that I will need to revert.

Anyway you are right, this work is not ready for review, I'll draft it.

**Comment by Arsnael** [comment]
```
[afe6e7e9b8b32498d069499071fd9e2a2636ed3a] [ERROR] Errors: 
[afe6e7e9b8b32498d069499071fd9e2a2636ed3a] [ERROR]   DomainUserMaildirMailboxManagerTest$BasicFeaturesTests>MailboxManagerTest$BasicFeaturesTests.renameMailboxShouldChangeTheMailboxPathOfTheChildMailbox:1582 » MailboxNotFound
```

**Comment by chibenwa** [comment]
Would it makes sense to extract `MailboxManager::renameMailboxShouldChangeTheMailboxPathOfTheChildMailbox` in a separate PR?

**Comment by chibenwa** [comment]
`CassandraRenameTest>Rename.testRenameITALY:67 » InvalidServerResponse `

**Comment by chibenwa** [comment]
That's a good occasion to strengthen a bit our MailboxManager test suite ;-)

**Comment by chibenwa** [comment]
`[5f591a8a07313454cba2134fc545350a60079c5c] [ERROR] src/main/java/org/apache/james/mailbox/store/MailboxReactorUtils.java:[40] (whitespace) EmptyLineSeparator: 'METHOD_DEF' should be separated from previous statement.`

**Comment by blackheaven** [comment]
squash and fix compilation

**Comment by chibenwa** [comment]
Thanks ;-)

**Comment by Arsnael** [comment]
```
[9d3fab55cae43b33680ee28b9cc1f43e19c95ba7] [ERROR] Failures: 
[9d3fab55cae43b33680ee28b9cc1f43e19c95ba7] [ERROR]   UserMailboxesRoutesTest$ExceptionHandling.deleteShouldReturnOkOnMailboxDoesNotExists:948 1 expectation failed.
[9d3fab55cae43b33680ee28b9cc1f43e19c95ba7] Expected status code <204> but was <500>.
```

**Comment by chibenwa** [comment]
```
[554d3cb50296d1e4749e0a37f44a47e4f2c65a68] [ERROR] A sharee should be able to retrieve a mailbox after sharer moved it into another mailbox  Time elapsed: 2.684 s  <<< FAILURE!
[554d3cb50296d1e4749e0a37f44a47e4f2c65a68] java.lang.AssertionError: 
[554d3cb50296d1e4749e0a37f44a47e4f2c65a68] 
[554d3cb50296d1e4749e0a37f44a47e4f2c65a68] Expecting:
[554d3cb50296d1e4749e0a37f44a47e4f2c65a68]   <[null]>
[554d3cb50296d1e4749e0a37f44a47e4f2c65a68] to contain only:
[554d3cb50296d1e4749e0a37f44a47e4f2c65a68]   <["2422f780-ab31-11ea-a3ad-ab532d52925b"]>
[554d3cb50296d1e4749e0a37f44a47e4f2c65a68] elements not found:
[554d3cb50296d1e4749e0a37f44a47e4f2c65a68]   <["2422f780-ab31-11ea-a3ad-ab532d52925b"]>
[554d3cb50296d1e4749e0a37f44a47e4f2c65a68] and elements not expected:
[554d3cb50296d1e4749e0a37f44a47e4f2c65a68]   <[null]>
[554d3cb50296d1e4749e0a37f44a47e4f2c65a68] 
[554d3cb50296d1e4749e0a37f44a47e4f2c65a68] 
[554d3cb50296d1e4749e0a37f44a47e4f2c65a68] [INFO] 
[554d3cb50296d1e4749e0a37f44a47e4f2c65a68] [INFO] Results:
[554d3cb50296d1e4749e0a37f44a47e4f2c65a68] [INFO] 
[554d3cb50296d1e4749e0a37f44a47e4f2c65a68] [ERROR] Failures: 
[554d3cb50296d1e4749e0a37f44a47e4f2c65a68] [ERROR]   RabbitMQMailboxSharingTest 
[554d3cb50296d1e4749e0a37f44a47e4f2c65a68] Expecting:
[554d3cb50296d1e4749e0a37f44a47e4f2c65a68]   <[null]>
[554d3cb50296d1e4749e0a37f44a47e4f2c65a68] to contain only:
[554d3cb50296d1e4749e0a37f44a47e4f2c65a68]   <["2422f780-ab31-11ea-a3ad-ab532d52925b"]>
[554d3cb50296d1e4749e0a37f44a47e4f2c65a68] elements not found:
[554d3cb50296d1e4749e0a37f44a47e4f2c65a68]   <["2422f780-ab31-11ea-a3ad-ab532d52925b"]>
[554d3cb50296d1e4749e0a37f44a47e4f2c65a68] and elements not expected:
[554d3cb50296d1e4749e0a37f44a47e4f2c65a68]   <[null]>
```

Now we are at the JMAP level <3

**Comment by blackheaven** [comment]
cannot reproduced

rebased to master

**Comment by Arsnael** [comment]
```
[9ba0c72e8f31fa5cf433eaef4f3fcea17b133009] [ERROR] Failures: 
[9ba0c72e8f31fa5cf433eaef4f3fcea17b133009] [ERROR]   FixingGhostMailboxTest.webadminCanMergeTwoMailboxesRights:268 1 expectation failed.
[9ba0c72e8f31fa5cf433eaef4f3fcea17b133009] JSON path [0][1].list[0].sharedWith doesn't match.
[9ba0c72e8f31fa5cf433eaef4f3fcea17b133009] Expected: map containing ["cedric@domain.tld"->ANYTHING]
[9ba0c72e8f31fa5cf433eaef4f3fcea17b133009]   Actual: {bob@domain.tld=[a, e, i, l, r, s, t, w]}
```
related or not?

**Comment by mbaechler** [comment]
I read the Cassandra part for now and it looks good.

If somebody is brave enough to read the other implementations, that would help going forward.

Otherwise I will do it in some times

**Comment by Arsnael** [comment]
Read it. Looks ok to me :)

**Comment by chibenwa** [comment]
Undrafted as this is :green_apple:

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraACLMapper.java [hunk 2]
```diff
@@ -120,21 +120,19 @@ private MailboxACL getAcl(CassandraId cassandraId, Row row) {
...
     }
 
-    public ACLDiff setACL(CassandraId cassandraId, MailboxACL mailboxACL) throws MailboxException {
+    public Mono<ACLDiff> setACL(CassandraId cassandraId, MailboxACL mailboxACL) {
         return updateAcl(cassandraId, acl -> new ACLWithVersion(acl.version, mailboxACL), mailboxACL)
             .flatMap(aclDiff -> userMailboxRightsDAO.update(cassandraId, aclDiff)
             .thenReturn(aclDiff))
-            .blockOptional()
-            .orElseThrow(() -> new MailboxException("Unable to update ACL"));
+            .switchIfEmpty(Mono.error(new MailboxException("Unable to update ACL")));
```
[the remark this answers is not in the document]
Comment (mbaechler) [reply mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraACLMapper.java:135]: (comment from last week, I just didn't send it)

[the remark this answers is not in the document]
Comment (chibenwa) [reply mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraACLMapper.java:135]: Let's defer that mono then...

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
```
RabbitMQDeletedMessageVaultIntegrationTest>DeletedMessageVaultIntegrationTest.vaultExportShouldExportZipContainsVaultMessagesToShareeWhenImapDeletedMailbox:543 
[9d129495ac63992dfd392d65e2a3188e7ac114c1] Expecting zipFile to contains 1 entries but actually contains (0) entries
```

Sounds unrelated. test this please

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageIdManager.java [hunk 3]
```diff
@@ -431,12 +433,12 @@ private void save(MessageIdMapper messageIdMapper, MailboxMessage mailboxMessage
 
 
     private Predicate<MailboxId> hasRightsOnMailbox(MailboxSession session, Right... rights) {
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageIdManager.java:435]: Hezy, that's clearly more work ;-)

**Comment by chibenwa** [comment]
:-( rebase conflict

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraACLMapper.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (mbaechler) [reply mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraACLMapper.java]: I don't get it, there's already a `Mono.error` that takes a supplier for that very reason (and I mentioned it). Why would we use defer instead?

[the remark this answers is not in the document]
Comment (chibenwa) [reply mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraACLMapper.java]: > (and I mentioned it).

Sorry I was unaware of it, sure I gonna use it.

**Comment by chibenwa** [comment]
test this please (build rescheduled)

**Comment by chibenwa** [comment]
404 test this please

**Comment by chibenwa** [comment]
```
[59d6906cc7c1d13023a4181dae7032bb4bd0a48c] 06:24:20.823 [ERROR] o.a.j.u.c.ConcurrentTestRunner - Error caught during concurrent testing (iteration 0, threadNumber 8)
[59d6906cc7c1d13023a4181dae7032bb4bd0a48c] org.apache.james.mailrepository.api.MailRepositoryStore$UnsupportedRepositoryStoreException: No Mail Repository associated with memory1
[59d6906cc7c1d13023a4181dae7032bb4bd0a48c] 	at org.apache.james.mailrepository.memory.MemoryMailRepositoryStore.lambda$retrieveMailRepository$2(MemoryMailRepositoryStore.java:153)
[59d6906cc7c1d13023a4181dae7032bb4bd0a48c] 	at java.base/java.util.Optional.orElseThrow(Optional.java:408)
[59d6906cc7c1d13023a4181dae7032bb4bd0a48c] 	at org.apache.james.mailrepository.memory.MemoryMailRepositoryStore.retrieveMailRepository(MemoryMailRepositoryStore.java:153)
[59d6906cc7c1d13023a4181dae7032bb4bd0a48c] 	at org.apache.james.mailrepository.memory.MemoryMailRepositoryStore.createNewMailRepository(MemoryMailRepositoryStore.java:115)
[59d6906cc7c1d13023a4181dae7032bb4bd0a48c] 	at com.github.fge.lambdas.functions.FunctionChainer.lambda$sneakyThrow$49(FunctionChainer.java:74)
[59d6906cc7c1d13023a4181dae7032bb4bd0a48c] 	at java.base/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1705)
[59d6906cc7c1d13023a4181dae7032bb4bd0a48c] 	at org.apache.james.mailrepository.memory.MemoryMailRepositoryStore.select(MemoryMailRepositoryStore.java:110)
[59d6906cc7c1d13023a4181dae7032bb4bd0a48c] 	at org.apache.james.mailrepository.memory.MemoryMailRepositoryStoreTest.lambda$selectShouldNotReturnDifferentResultsWhenUsedInAConcurrentEnvironment$1(MemoryMailRepositoryStoreTest.java:239)
[59d6906cc7c1d13023a4181dae7032bb4bd0a48c] 	at org.apache.james.util.concurrency.ConcurrentTestRunner$ConcurrentRunnableTask.run(ConcurrentTestRunner.java:175)
[59d6906cc7c1d13023a4181dae7032bb4bd0a48c] 	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
[59d6906cc7c1d13023a4181dae7032bb4bd0a48c] 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[59d6906cc7c1d13023a4181dae7032bb4bd0a48c] 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
[59d6906cc7c1d13023a4181dae7032bb4bd0a48c] 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
[59d6906cc7c1d13023a4181dae7032bb4bd0a48c] 	at java.base/java.lang.Thread.run(Thread.java:834)
```

???

**Comment by Arsnael** [comment]
```
[fdc10e867895771d0437195189bb7f2fca4623e5] [ERROR] Errors: 
[fdc10e867895771d0437195189bb7f2fca4623e5] [ERROR]   MemoryMailRepositoryTest.storingAndRemovingMessagesConcurrentlyShouldLeadToConsistentResult » NoClassDefFound
```

**Comment by chibenwa** [comment]
IMO we should drop this commit:

`Move MailRepositoryStore related classes to mailRepository-api `

It looks like an history management mistake (I'm sorry)

**Comment by Arsnael** [comment]
> IMO we should drop this commit:
> 
> `Move MailRepositoryStore related classes to mailRepository-api `
> 
> It looks like an history management mistake (I'm sorry)

My bad :)

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**